File: src/lib/controllers/Module.ts
import { open } from '@tauri-apps/api/dialog';
import { invoke } from "@tauri-apps/api";
import { reloadRepository } from "./Repository";

import type { Link } from '$lib/components/structs/Link';
import type { View } from '$lib/components/structs/View';
import type { Module} from "$lib/components/structs/Module"
import type { Object } from "$lib/components/structs/Object";
import type { Template } from "$lib/components/structs/Template";
import type { TreeItem } from "$lib/components/structs/Tree"
import type { Baseline } from '$lib/components/structs/Baseline';
import type { ModuleManifest } from '$lib/components/structs/ModuleManifest';

export async function createModule(manifesf: ModuleManifest, parent: TreeItem) {
    try {
        const module = await invoke('create_module', { man: manifesf, parent: parent });
        reloadRepository();
        return module as Module;
    } catch (err) {
        console.log(err);
    }
}

export async function readModule(module: TreeItem) {
    return readModuleFromPath(module.path);
}

export async function updateModuleManifest(module: TreeItem, manifest: ModuleManifest) {
    const mod = await invoke('update_module', { path: module.path, man: manifest });
    return mod as Module;
}

export async function deleteModule(module: TreeItem) {
    return invoke('delete_module', {path: module.path})
}

export async function readBaselinedModule(modulePath: string, version: string) {
    return invoke('read_baselined_module', { path: modulePath, version });
}

export async function readModuleFromPath(path: string) {
    return invoke('read_module', { path: path })
        .then((mod) => {
            return mod as Module;
        })
        .catch((e) => { 
            console.log(`Erro reading module at path ${path}: ${e}`) 
            return null;
        });
}

export function readBaselinedObject(path: string, id: number, version: string) {
    return invoke('read_object_from_baseline', {path, id, version});
 
}

export function readBaselinedObjects(path: string, version: string) {
    return invoke('read_objects_from_baseline', {path, version});
}

export async function createObject(modulePath: String, object: Object) {
    console.log(object)
    const mod = await invoke('create_object', { path: modulePath, object: object });
    return mod as Object;
}

export async function createDraftObject(modulePath: String, object: Object) {
    const mod = await invoke('create_draft_object', { path: modulePath, object: object as Object });
    return mod as Object;
}

export async function createLink(originModulePath: string, from: Link, to: Link) {
    return invoke('create_link', {originModulePath, from, to})
}

export async function deleteLink(originModulePath: string, from: Link, to: Link) {
    return invoke('delete_link', {originModulePath, from, to})
}

export async function readObjects(modulePath: String) {
    return invoke('read_objects', { path: modulePath });
}

export async function readDraftObjects(modulePath: String) {
    return invoke('read_draft_objects', { path: modulePath });
}

export async function readCurrentObjects(modulePath: string) {
    return invoke('read_current_objects', {path: modulePath});
}

export function deleteObject(modulePath: String, id: number) {
    return invoke('delete_object', {path: modulePath, id: id})
}

export function restoreObject(modulePath: String, id: number) {
	return invoke('restore_object', {path: modulePath, id: id})
}

export function updateTemplate(modulePath: string, template: Template) {
    return invoke('update_template', {path: modulePath, template: template})
}

export function updateViews(modulePath: string, views: View[]) {
    return invoke('update_views', {path: modulePath, views})
}

export function saveTemplate(modulePath: String, template: Template) {
    return invoke('update_template', {path: modulePath, template: template})
}

export async function createBaseline(modulePath: String, baseline: Baseline) {
    return invoke('create_baseline', {path: modulePath, baseline})
}

export async function exportCSV(modulePath: String) {
    const folder = await open({
        directory: true,
        multiple: false
    });
    if (folder) {
        let path = folder as string;
        return invoke('export_csv', {modulePath: modulePath, filePath: path})
    }
}

export async function exportXlsx(modulePath: String) {
    const folder = await open({
        directory: true,
        multiple: false
    });
    if (folder) {
        let path = folder as string;
        return invoke('export_xlsx', {modulePath: modulePath, filePath: path})
    }
}

File: src-tauri/src/handlers/module.rs
use std::{path::PathBuf, sync::Mutex};
use git2::Repository as GitRepository;
use tauri::{command, State}; 

use crate::core::{error::OpenDoorsError, module::{Baseline, Object, Template, Module, ModuleManifest}, tree::TreeItem};

#[command] 
pub fn create_module(state: State<'_, Mutex<Option<GitRepository>>>, man: ModuleManifest, parent: TreeItem) -> Result<Module, OpenDoorsError> {
	let repo = state.lock().unwrap();
	Ok(Module::create(&repo, &parent.path.into(), &man)?)
}

#[command]
pub fn read_module(path: PathBuf) -> Result<Module, OpenDoorsError> {
	Ok(Module::read(&path)?)
}

#[command]
pub fn update_module(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, man: ModuleManifest) -> Result<Module, OpenDoorsError> {
	let repo = state.lock().unwrap();
	Module::update(&repo, &path, &man)?;
	Ok(Module::read(&path)?)
}

#[tauri::command]
pub fn delete_module(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf) -> Result<(), String> {
	let repo = state.lock().unwrap();
  	Ok(())
}

#[command]
pub fn create_object(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, object: Object) -> Result<Object, OpenDoorsError> {
	let mut module = Module::read(&path)?;
	let repo = state.lock().unwrap();
	Ok(module.create_object(&repo, &mut object.clone())?)
}

#[command]
pub fn create_draft_object(path: PathBuf, object: Object) -> Result<Object, OpenDoorsError> {
	let mut module = Module::read(&path)?;
	Ok(module.create_draft_object(&mut &mut object.clone())?)
}

#[command]
pub fn read_object(path: PathBuf, id: usize) -> Result<Object, OpenDoorsError> {
	let module = Module::read(&path)?;
	Ok(module.read_object(id)?)
}

#[command]
pub fn read_draft_object(path: PathBuf, id: usize) -> Result<Object, OpenDoorsError> {
	let module = Module::read(&path)?;
	Ok(module.read_draft_object(id)?)
}

#[command]
pub fn read_objects(path: PathBuf) -> Result<Vec<Object>, OpenDoorsError> {
	let module = Module::read(&path)?;
	Ok(module.read_objects()?)
}

#[command]
pub fn read_draft_objects(path: PathBuf) -> Result<Vec<Object>, OpenDoorsError> {
	let mut module = Module::read(&path)?;
	Ok(module.read_draft_objects()?)
}

#[command] 
pub fn update_object(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, object: Object) -> Result<Object, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let mut module = Module::read(&path)?;
	let mut obj = object.clone();
	Ok(module.update_object(&repo, &mut obj)?) 
}

#[command] 
pub fn update_draft_object(path: PathBuf, object: Object) -> Result<Object, OpenDoorsError> {
	let mut module = Module::read(&path)?;
	let mut obj = object.clone();
	Ok(module.update_draft_object(&mut obj)?)
}

#[command] 
pub fn delete_object(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, id: usize) -> Result<Object, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let mut module = Module::read(&path)?;
	Ok(module.delete_object(&repo, id)?)
}

#[command] 
pub fn restore_object(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, id: usize) -> Result<Object, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let mut module = Module::read(&path)?;
	Ok(module.restore_object(&repo, id)?)
}

#[command]
pub fn create_template(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, template: Template) -> Result<Template, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let module = Module::read(&path)?;
	Ok(module.create_template(&repo, template)?)
}

#[command]
pub fn read_template(path: PathBuf) -> Result<Template, OpenDoorsError> {
	let module = Module::read(&path)?;
	Ok(module.read_template()?)
}

#[command]
pub fn update_template(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, template: Template) -> Result<Template, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let module = Module::read(&path)?;
	Ok(module.update_template(&repo, template)?)
}

#[command]
pub fn create_baseline(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, baseline: Baseline) -> Result<(), OpenDoorsError> {
	let repo = state.lock().unwrap();
	let mut module = Module::read(&path)?;
	module.create_baseline(&repo, &baseline.version.to_string(), &baseline.description)?;
	Ok(())
}

#[command]
pub fn read_object_from_baseline(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, id: usize, version: String) -> Result<Object, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let module = Module::read(&path.into())?;
	Ok(module.read_object_from_baseline(&repo, &id, &version)?)
}

#[command]
pub fn read_objects_from_baseline(state: State<'_, Mutex<Option<GitRepository>>>, path: PathBuf, version: String) -> Result<Vec<Object>, OpenDoorsError> {
	let repo = state.lock().unwrap();
	let module = Module::read(&path.into())?;
	Ok(module.read_objects_from_baseline(&repo, &version)?)
}
